src/core/agent/services/multiSigService.ts

```diff
@@ -716,6 +722,149 @@ class MultiSigService extends AgentService {
+      const roleims = d(
+        messagize(rpy, sigers, seal, undefined, undefined, false)
+      );
+      const atc = roleims.substring(rpy.size);
+      const roleEmbeds = {
+        rpy: [rpy, atc],
+      };
+
+      const recp = multisigMembers
+        .filter((signing: any) => signing.aid !== ourIdentifier.id)
+        .map((member: any) => member.aid);
+      const ourAid: Aid = await this.props.signifyClient
+        .identifiers()
+        .get(ourIdentifier.signifyName as string);
+      await this.sendMultisigExn(
+        ourIdentifier?.signifyName,
+        ourAid,
+        MultiSigRoute.RPY,
+        roleEmbeds,
+        recp,
+        { gid: aid }
+      );
+      results.push(op);
+    }
+    return results;
+  }
+
+  async joinAuthorization(notificationSaid: string) {
+    const exchangeMessage = await this.props.signifyClient
+      .exchanges()
+      .get(notificationSaid);
+    const multisigAid = exchangeMessage.exn.a.gid;
+    const multisigMetadataRecord =
+      await this.identifierStorage.getIdentifierMetadata(multisigAid);
+    const multisigSignifyName = multisigMetadataRecord.signifyName;
+    const { ourIdentifier, multisigMembers } =
+      await this.getMultisigParticipants(multisigSignifyName);
+    const icpMsg = await this.props.signifyClient
+      .groups()
+      .getRequest(notificationSaid);
+    if (!icpMsg.length) {
+      throw new Error(
+        `${MultiSigService.EXN_MESSAGE_NOT_FOUND} ${notificationSaid}`
+      );
+    }
+    const exn = icpMsg[0].exn;
+    // stamp, eid and role are provided in the exn message
+    const rpystamp = exn.e.rpy.dt;
+    const rpyrole = exn.e.rpy.a.role;
+    const rpyeid = exn.e.rpy.a.eid;
+    const endRoleRes = await this.props.signifyClient
+      .identifiers()
+      .addEndRole(multisigSignifyName, rpyrole, rpyeid, rpystamp);
+    const op = await endRoleRes.op();
+    const rpy = endRoleRes.serder;
+    const sigs = endRoleRes.sigs;
+
+    const hab = await this.props.signifyClient
+      .identifiers()
+      .get(multisigSignifyName);
+    const mstate = hab["state"];
+    const seal = [
+      "SealEvent",
+      { i: hab["prefix"], s: mstate["ee"]["s"], d: mstate["ee"]["d"] },
+    ];
+    const sigers = sigs.map((sig) => new Siger({ qb64: sig }));
+    const roleims = d(
+      messagize(rpy, sigers, seal, undefined, undefined, false)
+    );
+    const atc = roleims.substring(rpy.size);
+    const roleEmbeds = {
+      rpy: [rpy, atc],
+    };
+    const recp = multisigMembers
+      .filter((signing: any) => signing.aid !== ourIdentifier.id)
+      .map((member: any) => member.aid);
+    const ourAid: Aid = await this.props.signifyClient
+      .identifiers()
+      .get(ourIdentifier.signifyName as string);
+
+    await this.sendMultisigExn(
+      ourIdentifier.signifyName,
+      ourAid,
+      MultiSigRoute.RPY,
+      roleEmbeds,
+      recp,
+      { gid: hab["prefix"] }
+    );
+    return op;
+  }
```

Comment:
We need to use these functions somewhere. What is the flow? Once a multi-sig has been created and all have joined (not pending anymore), then one person adds the end role and the others join?

Reply:
I think we can trigger `endRoleAuthorization` when the muli-sig creating operation is done from the initiator's side. When another participant receives the notifications we can handle it and call `joinAuthorization` depending on our new notification flow.

Reply:
OK, but what about the case where it's a 2 of 5 multi-sig - so the operation is done once 2 people have signed - if the end role authorisation is sent out now, what will happen for the other 3 people?

Reply:
They can get the oobi if the threshold is only 2

Reply:
yes

Reply:
OK - and we can only start the end role authorisation after the multi-sig is not pending, or we can do it immediately?